core/amber/src/main/scala/edu/uci/ics/texera/web/resource/UserDictionaryResource.scala

```diff
@@ -0,0 +1,238 @@
+      Response
+        .status(Response.Status.BAD_REQUEST)
+        .entity(ErrorResponse("Malformed JSON payload"))
+        .build()
+    } else if (req.requestType == GetRequest.GET_ENTRY) {
+      if (!dictEntryExists(user, req.key))
+        Response.status(422).entity(ErrorResponse("No such entry")).build()
+      else Response.ok(EntryResponse(getValueByKey(user, req.key))).build()
+    } else {
+      Response.ok(DictResponse(getDict(user))).build()
+    }
+  }
+
+  /**
+    * This method creates or updates an entry in the current in-session user's dictionary based on
+    * the "key" and "value" attributes of the PostRequest
+    * @param session HttpSession
+    * @param req PostRequest
+    * @return Response, 200 ok - payload: ConfirmationResponse("Created" or "Updated"),
+    * 401 unauthorized - payload: ErrorResponse("Invalid session"),
+    * 400 bad request - payload: ErrorResponse("Malformed JSON payload"),
+    */
+  @POST
+  @Path("/set")
+  def setValue(
+      @Session session: HttpSession,
+      req: PostRequest
+  ): Response = {
+    val user = UserResource.getUser(session).orNull
+    if (user == null)
+      Response.status(Response.Status.UNAUTHORIZED).entity(ErrorResponse("Invalid session")).build()
+    else if (req == null || req.key == null || req.value == null) {
+      Response
+        .status(Response.Status.BAD_REQUEST)
+        .entity(ErrorResponse("Malformed JSON payload"))
+        .build()
+    } else if (dictEntryExists(user, req.key)) {
+      userDictionaryDao.update(new UserDictionary(user.getUid, req.key, req.value))
+      Response.ok(ConfirmationResponse("Updated")).build()
+    } else {
+      userDictionaryDao.insert(new UserDictionary(user.getUid, req.key, req.value))
+      Response.ok(ConfirmationResponse("Created")).build()
+    }
+  }
+
+  /**
+    * This method deletes a key-value pair from the current in-session user's dictionary based on
+    * the "key" attribute of the DeleteRequest
+    * @param session HttpSession
+    * @param req DeleteRequest
+    * @return Response, 200 ok - payload: "Deleted",
+    * 401 unauthorized - payload: "Invalid session",
+    * 400 bad request - payload: ErrorResponse("Malformed JSON payload"),
+    * 422 Unprocessable Entity - payload: "no such entry" (if no entry exists for provided key)
+    */
+  @DELETE
+  @Path("/delete")
+  def deleteValue(@Session session: HttpSession, req: DeleteRequest): Response = {
+    val user = UserResource.getUser(session).orNull
+    if (user == null)
+      Response.status(Response.Status.UNAUTHORIZED).entity(ErrorResponse("Invalid session")).build()
+    else if (req == null || req.key == null) {
+      Response
+        .status(Response.Status.BAD_REQUEST)
+        .entity(ErrorResponse("Malformed JSON payload"))
+        .build()
+    } else if (!dictEntryExists(user, req.key)) {
+      Response.status(422).entity(ErrorResponse("No such entry")).build()
+    } else {
```

Comment:
you don't need to check if it exists before deleting it. just execute the delete and let database to handle if it exist or not. In your way you are sending two queries to DB for each delete.

Reply:
[wont-fix]: I think it's important to throw an error when deleting a nonexistent entry. Even if it requires two queries. 

It's not elegant, but it shouldn't take that many resources compared to saving/retrieving other user resources

Reply:
With the single query you should be able to get an error if the entry to be deleted does not exist as well

Reply:
My bad, if it doesn't give error, it's fine to keep like this.

Reply:
I need this check to send an error response if the dict entry isn't there. I think it's worth a little inefficiency to have more verbose output.